docs/proposals/set-ref.md

```diff
@@ -0,0 +1,144 @@
+    Kind string
+    // Name of the referent.
+    Name string
+    // UID of the referent.
+    UID types.UID
+    // should this SetReference be treated as a parent for cascading deletion
+    isParent bool
+}
+```
+
+Most controllers (all that manage collections of things defined by label selector) will have slightly changed semantics: currently controller owns an object if its selector matches object’s labels, but after introduction of *setReference* a controller will own an object iff selector matches labels and *setReference* points to it.
+
+If the owner's selector or owned object's labels change, the owning controller will be responsible for orphaning (clearing *setReference*) objects, after which adoption procedure (setting *setReference*) might occur, if another controller has a selector matching..
+
+User shouldn't set *setReference* by himself/herself, only system components should be responsible for updating this field. In the fullness of time an Admission Control plugin may block all updates to *setReference* signed by something different than Controller Manager.
+
+For debugging purposes we want to add an `adoptionTime` annotation which will keep the time of last ownership transfer.
+
+# Upgrade/downgrade procedure
+
+During system upgrade controllers will adopt objects that match their label selector using ordinary adoption procedure. Adoption, by default, will set *IsParent* to *false,* so the behavior of objects existing before the update won’t change (i.e. deleting owners won't cause cascading deletion - see OwnerReference design).
+
+In case of system downgrade controllers will ignore *setReference* and everything will work as it does in the old version of the system.
+
+In case of subsequent system upgrade GarbageCollector will clear invalid *setReferences* and normal adoption will take place afterwards.
+
+# Orphaning/adoption
+
+Controllers will orphan objects they own in two cases:
+* Change of label/selector causing selector to stop matching labels (executed by the controller)
+* Deletion of a controller with *orphaning=true* (executed by the garbage collector)
+
+We will need a secondary orphaning mechanism in case of unclean controller deletion:
+* GarbageCollector will remove *setReference* from objects that no longer points to existing controllers
+
+Controller will adopt an object iff:
+* *setReference* is empty
+* *deletionTimestamp* is not set
+and
+* Controller is the oldest controller of a given kind from ones which has a selector that matches labels of adopted object
+
+Note that there are possible races during adoption if multiple *kinds* of controllers can own a given object (e.g. *Pod* can be owned by either a *Job* or a *ReplicaSet*).
+
+To prevent re-adoption of an object during deletion the *DeletionTimestamp* will be set when deletion is starting. When a controller has a non-nil *DeletionTimestamp* it won’t take any actions except updating its Status (in particular it won’t adopt any objects).
```

Comment:
I guess this is a comment to the paragraph about races between different kinds of controllers. The problem is that making is fully deterministic is impossible in current system design, as there can be third party controllers that are running in separate binary. To make it work we'd need to require all controllers to register somewhere (API server? ControllerController?) and this entity would need to make decisions about adoption.